nixos/modules/services/databases/postgresql.nix

```diff
@@ -92,14 +92,64 @@ in
         '';
       };
 
-      ensureDatabases = mkOption {
-        type = types.listOf types.str;
+      createDatabases = mkOption {
```

Comment:
@talyz I'm not a fan of how `createDatabases` is *not* reproducible, where `ensureDatabases` was.

Reply:
Hm, I'm not sure what you mean by that - in what way is it less reproducible than `ensureDatabases`?

Reply:
The name of the option `create` vs `ensure` already says half the story. If I modify the `owner` in `createDatabases` after I've already deployed once the database won't be created as described. Hence the rename to `create` makes sense. This is one reason why `ensureDatabases` is more appealing than `createDatabases`. Another reason being compatibility with `services.mysql.ensureDatabases`.

You'll have to forgive my lack of fluency in `postgresql`, so if what I'm saying is wrong please let me know why. What are the implications of the current implementation of `services.postgresql.ensureDatabases` as it stands if you were to use it in this module?

Reply:
The name was actually chosen because I wanted to implement it in a backwards compatible way, but if you would prefer it, I can keep the `ensureDatabases` name, since I think the backwards compatibility should be dropped anyway. One reason to keep the `createDatabases` name would, as you hint at, be to differentiate it to `services.mysql.ensureDatabases`; I don't see why the functionality of the postgres module should be held back to be compatible with the mysql module.

What I could do is to implement owner switching - I didn't really have time to research it before, but it seems doable. That should get it up to the same level of reproducibility as `ensureDatabases`.

Reply:
It's about more generalizing the database part in each nixos module, see the discussion starting at https://github.com/NixOS/nixpkgs/pull/56720#issuecomment-473255201.

Reply:
I've implemented owner switching now, so it should be more reproducible.

I don't really know how to address the concern of keeping the interface similar to the mysql one, and at the moment there doesn't seem to be an answer to how the generalized interface would work. I can imagine the following solutions:

1. The interfaces are allowed to differ and the general interface wraps them as necessary
2. The interfaces have to be identical and we add a new `ensureExtensions` option
3. The interfaces have to be identical and we change the mysql module to match the postgresql one
4. We don't touch the postgresql module and come up with a different solution (the original approach, for example)
5. Something else...?

To me, solution 1 and 3 are obviously preferable. Solution 2 would be pretty ugly - extensions are added at database level, not globally. I think it would be a shame not to have `createDatabases`, since it provides an easy way to define database properties that can trivially be extended in the future.

Reply:
This was resolved by moving the necessary stuff from the postgres module over to the gitlab module: we now inject what's needed into `systemd.services.postgresql.postStart` from there. The hope is that a better general database api will make this redundant by 20.03.